src/application.rs

```diff
@@ -82,11 +82,91 @@ pub trait ApplicationHandler<T: 'static = ()> {
+    /// there may be several executions of the event loop, including multiple redraws to windows,
+    /// between [`EventLoopProxy::wake_up()`] being called and the event being delivered.
+    ///
+    /// [`EventLoopProxy::wake_up()`]: crate::event_loop::EventLoopProxy::wake_up
+    ///
+    /// # Example
+    ///
+    /// Use a [`std::sync::mpsc`] channel to handle events from a different thread.
+    ///
+    /// ```no_run
+    /// use std::sync::mpsc;
+    /// use std::thread;
+    /// use std::time::Duration;
+    ///
+    /// use winit::application::ApplicationHandler;
+    /// use winit::event_loop::{ActiveEventLoop, EventLoop};
+    ///
+    /// struct MyApp {
+    ///     receiver: mpsc::Receiver<u64>,
+    /// }
+    ///
+    /// impl ApplicationHandler for MyApp {
+    ///     # fn window_event(
+    ///     #     &mut self,
+    ///     #     _event_loop: &ActiveEventLoop,
+    ///     #     _window_id: winit::window::WindowId,
+    ///     #     _event: winit::event::WindowEvent,
+    ///     # ) {
+    ///     # }
+    ///     #
+    ///     # fn resumed(&mut self, _event_loop: &ActiveEventLoop) {}
+    ///     #
+    ///     fn proxy_wake_up(&mut self, _event_loop: &ActiveEventLoop) {
+    ///         // Iterate current events, since wake-ups may have been merged.
+    ///         //
+    ///         // Note: We take care not to use `recv` or `iter` here, as those are blocking,
+    ///         // and that would be bad for performance and might lead to a deadlock.
+    ///         for i in self.receiver.try_iter() {
+    ///             println!("received: {i}");
+    ///         }
+    ///     }
+    ///
+    ///     // Rest of `ApplicationHandler`
+    /// }
+    ///
+    /// fn main() -> Result<(), Box<dyn std::error::Error>> {
+    ///     let event_loop = EventLoop::new()?;
+    ///     let proxy = event_loop.create_proxy();
+    ///
+    ///     let (sender, receiver) = mpsc::channel();
+    ///
+    ///     let mut app = MyApp { receiver };
+    ///
+    ///     thread::scope(|s| {
+    ///         // Send an event in a loop
+    ///         s.spawn(|| {
+    ///             let mut i = 0;
+    ///             loop {
+    ///                 println!("sending: {i}");
+    ///                 if sender.send(i).is_err() {
+    ///                     // Stop sending once `MyApp` is dropped
+    ///                     break;
+    ///                 }
+    ///                 // Trigger the wake-up _after_ we placed the event in the channel.
+    ///                 // Otherwise, `proxy_wake_up` might be triggered prematurely.
+    ///                 proxy.wake_up();
+    ///                 i += 1;
+    ///                 thread::sleep(Duration::from_secs(1));
+    ///             }
+    ///         });
+    ///
+    ///         Ok(event_loop.run_app(&mut app)?)
+    ///     })
+    /// }
+    /// ```
+    fn proxy_wake_up(&mut self, event_loop: &ActiveEventLoop) {
```

Comment:
Okay so... Actually, I think we should call this `custom_wake_up`, since I want to do https://github.com/rust-windowing/winit/pull/3424, and in that case the wake-up isn't strictly tied to `EventLoopProxy` any more.

Sorry for changing my mind so many times about this, I'll do the rename myself.

```suggestion
    fn custom_wake_up(&mut self, event_loop: &ActiveEventLoop) {
```

Reply:
Would you mind leaving it as `proxy_wake_up()` until we get to #3424?

Reply:
Sure